libs/core/synchronization/include/hpx/synchronization/range_mutex.hpp

```diff
@@ -25,89 +25,114 @@ namespace hpx::synchronization {
-          : lockRef(lock)
+          : mutex_ref(lock)
         {
-            lockId = lockRef.get().lock(begin, end);
+            lock_id = mutex_ref.get().lock(begin, end);
         }
         ~range_guard()
         {
-            lockRef.get().unlock(lockId);
+            mutex_ref.get().unlock(lock_id);
+        }
+
+        range_guard(range_guard<RangeMutex>&) = delete;
+        range_guard<RangeMutex>& operator=(range_guard<RangeMutex>&) = delete;
+
+        range_guard(range_guard<RangeMutex>&& rhs_lock)
+        {
+            mutex_ref.get().unlock(lock_id);
+            mutex_ref = rhs_lock.mutex_ref;
+            lock_id = rhs_lock.lock_id;
+            rhs_lock.lock_id = 0;
+        }
+
+        range_guard<RangeMutex>& operator=(range_guard<RangeMutex>&& rhs_lock)
+        {
+            mutex_ref.get().unlock(lock_id);
+            mutex_ref = rhs_lock.mutex_ref;
+            lock_id = rhs_lock.lock_id;
+            rhs_lock.lock_id = 0;    // invalidating rhs_lock
         }
     };
 
     template <typename RangeMutex>
     class range_unique_lock
     {
-        std::reference_wrapper<RangeMutex> lockRef;
-        std::size_t lockId = 0;
+        std::reference_wrapper<RangeMutex> mutex_ref;
+        std::size_t lock_id = 0;
 
     public:
         range_unique_lock(RangeMutex& lock, std::size_t begin, std::size_t end)
-          : lockRef(lock)
+          : mutex_ref(lock)
         {
-            lockId = lockRef.get().lock(begin, end);
+            lock_id = mutex_ref.get().lock(begin, end);
         }
 
         ~range_unique_lock()
         {
-            lockRef.get().unlock(lockId);
+            mutex_ref.get().unlock(lock_id);
         }
 
-        void operator=(range_unique_lock<RangeMutex>&& lock)
+        range_unique_lock(range_guard<RangeMutex>&) = delete;
+        range_unique_lock<RangeMutex>& operator=(
+            range_unique_lock<RangeMutex>) = delete;
+
+        range_unique_lock<RangeMutex>& operator=(
+            range_unique_lock<RangeMutex>&& rhs_lock)
         {
-            lockRef.get().unlock(lockId);
-            lockRef = lock.lockRef;
-            lockId = lock.lockRef.get().lock();
+            mutex_ref.get().unlock(lock_id);
+            mutex_ref = rhs_lock.mutex_ref;
+            lock_id = rhs_lock.lock_id;
+            rhs_lock.lock_id = 0;    // invalidating rhs_lock
         }
 
         void lock(std::size_t begin, std::size_t end)
         {
-            lockId = lockRef.get().lock(begin, end);
+            lock_id = mutex_ref.get().lock(begin, end);
         }
 
         void try_lock(std::size_t begin, std::size_t end)
         {
-            lockId = lockRef.get().try_lock(begin, end);
+            lock_id = mutex_ref.get().try_lock(begin, end);
         }
 
         void unlock()
         {
-            lockRef.get().unlock(lockId);
-            lockId = 0;
+            mutex_ref.get().unlock(lock_id);
+            lock_id = 0;
```

Comment:
Shouldn't this be made thread-safe?

Reply:
@hkaiser aren't lock guards supposed to be thread local usually? I am unsure how lock_id can be racing if lock_guards are local.

Reply:
You are certainly right, my fault.